### ISSUE TITLE
Consider AEGIS-128X

### ISSUE THREAD

**Issue by jedisct1** [jedisct1/draft-aegis-aead#75]
https://github.com/jedisct1/aegis-128X

**Comment by jedisct1** [comment]
Specification, benchmarks and rationale: https://eprint.iacr.org/2023/523

**Comment by emanjon** [comment]
I think this would be an excellent option for mutually authenticated TLS in infrastructure. TLS, DTLS, and QUIC are heavily used to protect signaling between nodes in e.g. the 5G core. The performance gains are impressive. I support inclusion of 128X2, 128X4, 256X2, 256X4

**Comment by jedisct1** [comment]
Even if only recent Intel CPUs can benefit from it, I also think it would be worth having in the portfolio.

Let's tag a new version of the draft, which addresses the issues reported here and on the list, and we can then focus on adding the parallel versions in the next one.

**Comment by jedisct1** [comment]
For usage with TLS, that will require 4 new identifiers from IANA. That's quite a lot just for AEGIS. Wondering if that can be an issue.

**Comment by emanjon** [comment]
Both the AEAD and TLS cipher suite have 65536 values so technically the IANA part is not a problem. But people might not like it. You also need to put all the cipher suites in the TLS negotiation.

**Comment by samuel-lucas6** [comment]
To play devil's advocate:

1. How much longer/more complicated would this make the draft? It's very readable currently.
2. Is the CFRG process likely to be slowed because of this? I don't know the details of why XChaCha20/XChaCha20-Poly1305 never proceeded, but obviously none of us want that to happen with AEGIS.
3. Aren't people likely to be confused about when to use which? Some people may also interpret the X to mean eXtended like XChaCha20.
4. Will people want to use/support the faster variants any time soon given they're only for newer CPUs? I don't know how this would work in the real world unless they're doing the AES-GCM except when it's not supported so switch to ChaCha20-Poly1305 approach. Then in terms of APIs/protocols, I don't think I'd want to support 4 different variants. Even 2 variants is somewhat annoying, which is where Rocca-S has an advantage.

It's a tricky one though because it also does make sense to have them in the same document given they're 'modes', and the performance is obviously excellent future-proofing.

**Comment by jedisct1** [comment]
Adding the X variants shouldn't complicate the specification, especially if we simply define them in terms of 128L and 256. If "X" is confusing, that can still be renamed ("P" for "parallelism"? But `X2` and `X4` seemed quite intuitive to me).

Some options:

1) Only specify `128X2` and `256X2`. That's what the vast majority of servers support today. CPUs with 4-way VAES are quite new, and I don't even know of any cloud providers that supports them. Google Cloud apparently has an instance type that does, but it's currently only for testing purposes. I wasn't able to run benchmarks on such CPUs, so maybe the performance gains are even not as good as we may think.
2) Specify the parallel variants in a distinct document. That doesn't prevent them from being implemented, and eventually submitted to CFRG later.
3) On a CPU without VAES, the overhead of `128X2` over `128L` is an extra call to the initialization and finalization functions. The overhead of `128X4` is 3 extra calls to these functions. If the messages are large enough, that can be amortized. So, we could define a single `128X` variant, that is `128X1` for small messages, `128X2` for larger messages and `128X4` for even larger messages. Of course, we need to exactly define what these sizes are. This is similar to what e.g. KangarooTwelve is doing. The flip side is that CPUs with VAES don't get any performance boost for short messages.

**Comment by lin72h** [comment]
@jedisct1  Hi Frank, I totally support to bring AEGIS-128X in, since it make good use of VAES, instruction set.

I've been researching the CPU based LLM implementation, like [llama.cpp](https://github.com/ggerganov/llama.cpp).

I think more and more people are willing to upgrade their CPU to have VNNI or AMX instruction set to run these LLM model local efficiently. And good news is that most CPU which support VNNI like Zen3 or Intel Gen11 or Gen12 also comes with VAES too.

So VAES will not be a niche instruction set in the coming month I think. Let's make AEGIS-128X blazingly fast !

**Comment by jedisct1** [comment]
A good way to get more support for AEGIS-X would be to make it part of the Supercop benchmark.

But that doesn't solve the problem of having multiple variants.

**Comment by samuel-lucas6** [comment]
> If "X" is confusing, that can still be renamed ("P" for "parallelism"? But X2 and X4 seemed quite intuitive to me).

I think AEGIS-X sounds best; that was just a wild hypothesis (XAEGIS doesn't work). AEGIS-128V was my initial thought due to VAES, but it's less memorable/worse to say.

> Only specify 128X2 and 256X2. That's what the vast majority of servers support today.

That sounds sensible given what you're saying.

> Specify the parallel variants in a distinct document. That doesn't prevent them from being implemented, and eventually submitted to CFRG later.

Whilst this keeps the documents short, it may be harmful for adoption, like BLAKE2 and then BLAKE2X. I trust your judgement on the draft remaining readable. I'm just trying to be cautious/protective of what's been created because it really is well written for implementers.

> The flip side is that CPUs with VAES don't get any performance boost for short messages.

Tricky. That doesn't sound ideal. How short are we talking? Seems like stream encryption wouldn't reach `128X4`.

**Comment by jedisct1** [comment]
I finally got access to a CPU with 4-way VAES and AVX512 (AMD Ryzen 7 7700).

Here are the results for 16 Kb messages:

| Cipher | Throughput (Gb/s) |
|---|---|
| AES128-GCM | 38 |
| AES128-OCB | 79 |
| AEGIS-128L | 152 |
| AEGIS-128X2 | 310 |
| AEGIS-128X4 | 349 |


| Cipher | Throughput (Gb/s) |
|---|---|
| AES256-GCM | 36 |
| AES256-OCB | 56 |
| AEGIS-256 | 94 |
| AEGIS-256X2 | 184 |
| AEGIS-256X2 | 277 |

**Comment by jedisct1** [comment]
> Tricky. That doesn't sound ideal. How short are we talking? Seems like stream encryption wouldn't reach 128X4.

Just a back of the envelope computation, but we're talking 8K to reach 2x, 32K to reach 4x.

Not a big fan of this approach either, especially since it complicates implementations.

A simpler approach is to give a set of very simple rules for cipher selection in a TLS-like protocol.

For AEGIS-128:
* If the CPU has a 4-way AES instruction: include 128X4 in the cipher suite.
* If the CPU has a 2-way AES instruction: include 128X4 in the cipher suite.
* Always include 128L

Ditto for the 256 variants.

**Comment by lin72h** [comment]
amazing, I think if there's a way to output 256b Hash, AEGIX-128X could replace blake3 which use the tree based immediate 1KB chunk pollute all the cacheline in all the core and not achieve the same throughput like AEGIX-128X

**Comment by jedisct1** [comment]
@lin72h AEGIS is not a hash function, though. If the key is public or reusable, an attacker can compute inputs that will zero the state, the only defense left being the input length.

It's fine to use as a replacement for a hash function only if:
- There's a secret key
- Or the input is not attacker-controlled
- Or the key (or IV) are public, but unique to each message. Such output can be verified but will not be deterministic.

**Comment by jedisct1** [comment]
@lin72h But, by the way, yes, AEGIS-128L and X can already output a 256 bit tag.

**Comment by lin72h** [comment]
Thanks for the detailed explanation. I don't understand the details you mentioned fully, but I think for data integrity  check use case like ZFS's 256bit checksum,  I think it's very suitable than the blake3.

**Comment by samuel-lucas6** [comment]
Thanks for the benchmarks. It seems worth specifying x4.

> Ditto for the 256 variants. Don't bother complicating things with input sizes.

Ok we can rule that out then.

> Let's tag a new version of the draft, which addresses the issues reported here and on the list, and we can then focus on adding the parallel versions in the next one.

I'll try and tidy the references (#72) in a PR this weekend.

**Comment by jedisct1** [comment]
> Thanks for the benchmarks. It seems worth specifying x4.

I'm still a little bit on the fence. The speedup is not as significant as between x1 and x2. 

Looking at the graphs, it also takes more time for x4 to ramp up, presumably because of the CPU cache prefetching algorithm.

I'm gonna write a multi-thread benchmark, that encrypts various message sizes in parallel to better mimic concurrent TLS-like sessions. That may be more revealing than a micro benchmark.

**Comment by jedisct1** [comment]
Closing in favor of an issue about the actual addition of these variants to the document.
